### PR TITLE
refactor McSession

### PR DESCRIPTION
**Pull request by the-daniel-rothig**
### Context

https://trello.com/c/GZWlOaG8/272-refactor-getjsonuserdetailsfromdatabase-access-token-caching
and
https://trello.com/c/DpL48FW3/271-mcsession-duplicates-subjectid

### Changes proposed in this pull request

The original implementation contained a number of redundant steps because the caching was trying to replicate too much oauth behaviour. In fact, this had caused an eleventh-hour bug that broke Access Requests by overwriting the names of users when they re-authenticated!

Two changes:
- Remove the redundant subjectId 
- Remove behaviour that attempts to update user records when the cache is hit
- generally tidy up the code a bit.

### Guidance to review

No new tests as this is almost a pure refactor with no behaviour changes. Main change is that data updates now only happen if oAuth is hit, not when the cache is hit.

### CODE REVIEW

**Comment by the-daniel-rothig** [comment]
[mc_session duplicates subjectId](https://trello.com/c/DpL48FW3/271-mcsession-duplicates-subjectid)

**Comment by the-daniel-rothig** [comment]
[refactor GetJsonUserDetailsFromDatabase - access token caching](https://trello.com/c/GZWlOaG8/272-refactor-getjsonuserdetailsfromdatabase-access-token-caching)

File: src/ManageCourses.Api/Middleware/BearerTokenHandler.cs [hunk 1]
```diff
@@ -44,57 +44,56 @@ protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
...
             {
-                var userDetails = GetJsonUserDetailsFromDatabase(accessToken) ?? GetJsonUserDetailsFromOauth(accessToken);
-                
-                try
+                subjectAndEmail = GetSubjectAndEmailFromDatabase(accessToken);
+                if (subjectAndEmail == null)
                 {
+                    var userDetails = GetJsonUserDetailsFromOauth(accessToken);
+                    subjectAndEmail = new SubjectAndEmail(userDetails.Subject, userDetails.Email);
                     await _userService.UserSignedInAsync(accessToken, userDetails);
```
Comment (timabell) [inline src/ManageCourses.Api/Middleware/BearerTokenHandler.cs]: you'll lose the login log by moving the call to `UserSignedInAsync()` into the `if`

https://github.com/DFE-Digital/manage-courses-api/blob/ae1424fff491e74b8c7f002917f4da859f594cfe/src/ManageCourses.Api/Services/Users/UserService.cs#L46

reject

File: src/ManageCourses.Api/Middleware/BearerTokenHandler.cs [hunk 2]
```diff
@@ -44,57 +44,56 @@ protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
...
         {
             var dateCutoff = DateTime.UtcNow.AddMinutes(-30);
             var session = _manageCoursesDbContext.McSessions
                 .Include(x => x.McUser)
                 .Where(x => x.AccessToken == accessToken && x.CreatedUtc > dateCutoff)
-                .SingleOrDefault();
+                .FirstOrDefault(); /*   There is an edge case where more than one record for a valid session 
+                                        could be added, e.g. when clock skew occurs between two authentications. 
+                                        Any of these overlapping records would qualify and the redundancy is quite harmless. 
+                                        So we use First, not Single, here. */
```
Comment (timabell) [inline src/ManageCourses.Api/Middleware/BearerTokenHandler.cs]: :+1:

File: src/ManageCourses.Api/Middleware/BearerTokenHandler.cs (hunk 1 above)
Comment (the-daniel-rothig) [reply src/ManageCourses.Api/Middleware/BearerTokenHandler.cs]: fixed :+1:

**Comment by timabell** [comment]
Tests passing. I'm pretty happy with this structure. The tests could probably be more thorough but timeboxing for now.

**Pull request merged**